openedx/core/djangoapps/user_api/preferences/tests/test_api.py

```diff
@@ -0,0 +1,324 @@
+    are not specified.
+    """
+    password = "test"
+
+    def setUp(self):
+        super(TestPreferenceAPI, self).setUp()
+        self.user = UserFactory.create(password=self.password)
+        self.different_user = UserFactory.create(password=self.password)
+        self.staff_user = UserFactory(is_staff=True, password=self.password)
+        self.no_such_user = UserFactory.create(password=self.password)
+        self.no_such_user.username = "no_such_user"
+        self.test_preference_key = "test_key"
+        self.test_preference_value = "test_value"
+        UserPreference.set_preference(self.user, self.test_preference_key, self.test_preference_value)
+
+    def test_get_user_preference(self):
+        """
+        Verifies the basic behavior of get_user_preference.
+        """
+        self.assertEqual(
+            get_user_preference(self.user, self.test_preference_key),
+            self.test_preference_value
+        )
+        self.assertEqual(
+            get_user_preference(self.staff_user, self.test_preference_key, username=self.user.username),
+            self.test_preference_value
+        )
+
+    def test_get_user_preference_errors(self):
+        """
+        Verifies that get_user_preference returns appropriate errors.
+        """
+        with self.assertRaises(UserNotFound):
+            get_user_preference(self.user, self.test_preference_key, username="no_such_user")
+
+        with self.assertRaises(UserNotFound):
+            get_user_preference(self.no_such_user, self.test_preference_key)
+
+        with self.assertRaises(UserNotAuthorized):
+            get_user_preference(self.different_user, self.test_preference_key, username=self.user.username)
+
+    def test_get_user_preferences(self):
+        """
+        Verifies the basic behavior of get_user_preferences.
+        """
+        expected_user_preferences = {
+            self.test_preference_key: self.test_preference_value,
+        }
+        self.assertEqual(get_user_preferences(self.user), expected_user_preferences)
+        self.assertEqual(get_user_preferences(self.staff_user, username=self.user.username), expected_user_preferences)
+
+    def test_get_user_preferences_errors(self):
+        """
+        Verifies that get_user_preferences returns appropriate errors.
+        """
+        with self.assertRaises(UserNotFound):
+            get_user_preferences(self.user, username="no_such_user")
+
+        with self.assertRaises(UserNotFound):
+            get_user_preferences(self.no_such_user)
+
+        with self.assertRaises(UserNotAuthorized):
+            get_user_preferences(self.different_user, username=self.user.username)
+
+    def test_set_user_preference(self):
+        """
+        Verifies the basic behavior of set_user_preference.
+        """
+        set_user_preference(self.user, self.test_preference_key, "new_value")
+        self.assertEqual(
```

Comment:
I had that but it seemed odd because it only ever accepts its own username. I guess I'll add it back...